# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv](https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive-beta.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/mbc/ai/m4/datasets/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I would expect the Random Forest Classifer to perform a bit better. Page 4.4.3 demonstrated this possibility when we classified apps as malware. Given the nature of ensemble learning models, it would make sense that they'd be resistant to overfitting.

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
X = data.iloc[:,:-1]
y = data["spam"]

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

0    2788
1    1813
Name: spam, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train,X_test,y_train,y_test = train_test_split(X,y)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
SS = StandardScaler()

In [13]:
# Fit the Standard Scaler with the training data
scaled_data_train = SS.fit_transform(X_train)
scaled_data_test = SS.fit_transform(X_test)

In [14]:
# Scale the training data
scaled_data_train = pd.DataFrame(scaled_data_train, columns = X.columns)
scaled_data_test = pd.DataFrame(scaled_data_test, columns = X.columns)                           

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [15]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
LR = LogisticRegression(random_state=1)
LR.fit(scaled_data_train,y_train)
print(f"Logistic Regression Model Score: {LR.score(scaled_data_train,y_train)}")

Logistic Regression Model Score: 0.9284057971014493


In [19]:
# Make and save testing predictions with the saved logistic regression model using the test data
y_pred_test_LR = LR.predict(scaled_data_test)
# Review the predictions
y_pred_test_LR

array([1, 1, 1, ..., 1, 1, 0], dtype=int64)

In [20]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_pred_test_LR,y_test)

0.9304952215464813

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [24]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
RFC = RandomForestClassifier(random_state=1)
RFC.fit(scaled_data_train,y_train)
print(f"Random Forest Model Score: {RFC.score(scaled_data_train,y_train)}")

Random Forest Model Score: 0.9997101449275362


In [21]:
# Make and save testing predictions with the saved logistic regression model using the test data
y_pred_test_RFC = RFC.predict(scaled_data_test)

# Review the predictions
y_pred_test_RFC

array([1, 1, 1, ..., 1, 1, 0], dtype=int64)

In [22]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_pred_test_RFC,y_test)

0.946133796698523

In [23]:
accuracy_score(y_pred_test_RFC,y_test) - accuracy_score(y_pred_test_LR,y_test)

0.0156385751520417

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forest Classifier model performed a little bit better than the Logistic Regression model, in line with expectations. However, what is rather ironic is that the RFC model appears to have overfitted by training data judging by the 99.9% accuracy score versus 92.8% for the LR model. The LR scored better on the testing data than the training whereas the RFC scored better on the training data.